.NET/Microsoft.Recognizer.Text.TimeZone/Extractors/BaseTimeZoneExtractor.cs

```diff
@@ -0,0 +1,86 @@
+﻿using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Text.RegularExpressions;
+
+using Microsoft.Recognizers.Definitions;
+
+namespace Microsoft.Recognizers.Text.TimeZone
+{
+    public abstract class BaseTimeZoneExtractor : IExtractor
+    {
+        internal abstract ImmutableDictionary<Regex, string> Regexes { get; }
+
+        protected virtual string ExtractType { get; } = "";
+
+        public virtual List<ExtractResult> Extract(string source)
+        {
+            if (string.IsNullOrEmpty(source))
+            {
+                return new List<ExtractResult>();
+            }
+
+            var result = new List<ExtractResult>();
+            var matchSource = new Dictionary<Match, string>();
+            var matched = new bool[source.Length];
+            for (var i = 0; i < source.Length; i++)
+            {
+                matched[i] = false;
+            }
+
+            var collections = Regexes.ToDictionary(o => o.Key.Matches(source), p => p.Value);
+            foreach (var collection in collections)
+            {
+                foreach (Match m in collection.Key)
+                {
+                    for (var j = 0; j < m.Length; j++)
+                    {
+                        matched[m.Index + j] = true;
+                    }
+
+                    //Keep Source Data for extra information
+                    matchSource.Add(m, collection.Value);
+                }
+            }
+
+            var last = -1;
+            for (var i = 0; i < source.Length; i++)
+            {
+                if (matched[i])
+                {
+                    if (i + 1 == source.Length || !matched[i + 1])
+                    {
+                        var start = last + 1;
+                        var length = i - last;
+                        var substr = source.Substring(start, length);
+
+                        if (matchSource.Keys.Any(o => o.Index == start && o.Length == length))
+                        {
+                            var srcMatch = matchSource.Keys.First(o => o.Index == start && o.Length == length);
+                            var er = new ExtractResult
+                            {
+                                Start = start,
+                                Length = length,
+                                Text = substr,
+                                Type = ExtractType,
+                                Data = matchSource.ContainsKey(srcMatch) ? matchSource[srcMatch] : null
+                            };
+                            result.Add(er);
+                        }
+                    }
+                }
+                else
+                {
+                    last = i;
+                }
+            }
+
+            return result;
+        }
+    }
+    public enum TimeZoneMode
+    {
+        //Default is for unit and datetime
```

Comment:
I'm not sure what the comments means.
Also, if users of the timezone entity extractor code will need this enum, please move it to its own file.